docs/articles/api-layers.md

```diff
@@ -1,3 +1,130 @@
+Where gRPC access to an API is available, our client libraries use it.
+
+## Layers of gRPC-based libraries
+
+All gRPC-based libraries contain generated code, and some also contain hand-written code. The 
+generated code comes from three sources:
+
+- [Google API Tools](https://github.com/googleapis/toolkit) generation for clients
+- [gRPC](http://www.grpc.io) raw RPC service generation
+- [Protocol Buffer](https://developers.google.com/protocol-buffers/) generation
+  for request, response and resource messages and enums
+
+To show the layering involved, we'll use the Google.Datastore.V1Beta3 client library as an example.
+An application using this library will typically end up with the following assemblies:
+
+- Google.Datastore.V1Beta3: the client library itself
+- Google.Api.Gax: a support library providing functionality required across multiple
+  gRPC-based wrapper libraries, such as [page streaming](page-streaming.md).
+- Google.Protobuf: Protocol Buffer support library
+- Grpc.Core and Grpc.Core: gRPC support libraries
+- Google.Apis.Auth and Google.Apis.Auth.PlatformServices: support libraries used by both
+  REST service libraries and gRPC libraries for authentication
+- Dependencies from the above libraries for HTTP, JSON serialization, cryptography, and
+  asynchronous sequences
+
+Within `Google.Datastore.V1Beta3` there are various kinds of code, all within
+the `Google.Datastore.V1Beta3` namespace:
+
+- [DatastoreDb](../obj/api/Google.Datastore.V1Beta3.DatastoreDb.yml) etc: hand-written code
+  to support a higher layer of abstraction than the generated code. This provides a more
+  idiomatic experience than the autogenerated code, with consideration for things like
+  making [DatastoreTransaction](../obj/api/Google.Datastore.V1Beta3.DatastoreTransaction.yml)
+  implement `IDisposable`.
+- [DatastoreClient](../obj/api/Google.Datastore.V1Beta3.DatastoreClient.yml) is generated by the
+  Google API Tools. This is still a higher level of abstraction than the raw RPC, supporting 
+  features such as [page streaming](page-streaming.md) and automatic creation of requests from the 
+  constituent parameters, but it does not provide the level of idiomatic usage that the hand-written
+  abstraction does. This is available from a `DatastoreDb` via the
+  [DatastoreDb.Client](../obj/api/Google.Datastore.V1Beta3.DatastoreDb.yml#Google_Datastore_V1Beta3_DatastoreDb_Client) property if necessary.
+- [Datastore.DatastoreClient](../obj/api/Google.Datastore.V1Beta3.Datastore.DatastoreClient.yml) is
+  the gRPC-generated RPC access class. Very few applications will need to use this, but it is available from the higher level `DatastoreClient` class via the
+  [DatastoreClient.GrpcClient](../obj/api/Google.Datastore.V1Beta3.DatastoreClient.yml#Google_Datastore_V1Beta3_DatastoreClient_GrpcClient) property if necessary.
+- Additional hand-written code in the form of extension methods and addition to the partial 
```

Comment:
I wonder if this bullet should come second in this list? 
This list is from higher-level to lower-level code, except for this last point.